README-new.txt

```diff
@@ -0,0 +1,209 @@
+   The C is not meant to be read or maintained by humans, as it is a representation
+   of a low level IR (intermediate representation).
+    m3-sys/m3back/src/M3C.m3
+
+ -  A working LLVM backend?
+
+With some exception, the backends should generate code compatible with each other.
+But there is at least one exception so this is best avoided.
+Specifically related to nested functions or at least pointers to them.
+
+= The libraries
+
+The system includes a variety of libraries including but not limited to:
+  - threading and locking
+  - formated textual io (similar to printf)
+  - networking (sockets)
+  - custom GUI library ("Trestle")
+
+The system also includes "bindings" so you can directly call Win32, OpenGL, X Windows, ODBC, etc.
+ That is, C headers rewritten in Modula-3. Very much like P/Invoke in C#.
+
+As well some sample applications such as a web browser, forms editor, and some games.
+
+= Build system
+
+The compiler is also integrated into its own build system.
+
+The build system is mostly declarative, embedded in a somewhat general "scripting" language,
+ known as "quake" (quick make?)
+
+Architecturally it resembles scons, though the again the build system and compiler are linked.
+
+The build system contains per-platform "configuration" files, written in the underlying quake
+scripting language, providing functions like "compile a .c file", "link an executable"
+and "link a shared library". In this respect the build system largely subsumes libtool.
+ See m3-sys/cminstall/src/config-no-install
+
+= Building
+
+Because the system is written in Modula-3 building it from source can be challenging.
+
+Several distribution formats are provided.
+
+ - "boot" is generated C that produces the compiler, integrated and C backends, and build sytem.
+   These live in one executable -- "cm3".
+   "boot" also produces mklib which is used on Windows platforms only.
+   The C code is almost portable to all systems, however for now it is platform-specific.
+   Specifically it encodes sizeof(void*), endian, and Windows vs. Unix.
+   In time this matrix should be reduced.
+
+ - "min" is a minimal binary distribution containing little more than "boot" -- cm3 and mklib --
+   but also two libraries, m3core and libm3.
+     See m3-libs/m3core and m3-libs/libm3.
+   m3core is the lowest level library containing threading, garbage collector, unistd/win32 bindings.
+   The compiler and m3core are somewhat tightly coupled. The compiler outputs calls into m3core,
+   such as for garbage collection barriers.
+   libm3 is a higher level utility library containing things like growable vectors, linked lists, etc.
+
+ - "all" is everything. Sometimes some things are omitted, such as if a system lacked X Windows installed.
+
+ The use-cases are fairly obvious.
+   "min" is small and binary. You can write small Modula-3 programs right away.
+   "all" is larger and provides a larger assortment of libraries.
```

Comment:
Maybe throw in vernacular like "kitchen sink" or "batteries included", "other systems call it that".

Reply:
> - "all" is everything, batteries included

to be followed by a list.  I can try to put together that list.

Reply:
I am not sure the list is important, but to allude to it already being mentioned, "gui library", "browsers (not very good ones, but demos", "games (tetris)", Obliq, Mentor...maybe with links to papers .... presenting hierarhical information is always an exercise in how deep to go though... links certainly helps, I would not detail obliq and mentor but maybe link. They are kinda cool.

It is tough, you know, the system, the "installers", and the documentation, tries to be all things to all people, just what one wants. Providing sockets+stdio does satisfy a large group, but not everyone. You need gui to capture some people. etc. The group satisifed with sockets+stdio might complain about the large size of all.. no great answer, but I sorta tried, or rather I mostly copied what Olaf did, possibly misunderstanding it, and doing less (I am not sure he had overlap, and he had other groups like std, I tried to reduce things for a simpler story, but ultimately if you take a Debian-like perspective, what I did is quite bad..there should be no overlap....)